### ISSUE TITLE
mac128: infinite loop: execute_timers: timer callback floppy_image_device::index_resync

### ISSUE THREAD

**Issue by Davidian1024** [mamedev/mame#11585]
### MAME version

0.259 (mame0259-37-g6dea864283c)

### System information

Ubuntu 20.04.6, 64-bit
x86-64, 16GB system RAM, GeForce GTX 1080 Ti (rev a1), driver version 535.113.01

### INI configuration details

```INI
﻿#
# CORE CONFIGURATION OPTIONS
#
readconfig                1
writeconfig               0

#
# CORE SEARCH PATH OPTIONS
#
homepath                  .
rompath                   roms;/mnt/storage/ROMs/MAME
hashpath                  hash
samplepath                samples
artpath                   artwork
ctrlrpath                 ctrlr
inipath                   $HOME/.mame;.;ini
fontpath                  .
cheatpath                 cheat
crosshairpath             crosshair
pluginspath               plugins
languagepath              language
swpath                    software

#
# CORE OUTPUT DIRECTORY OPTIONS
#
cfg_directory             cfg
nvram_directory           nvram
input_directory           inp
state_directory           sta
snapshot_directory        snap
diff_directory            diff
comment_directory         comments
share_directory           share

#
# CORE STATE/PLAYBACK OPTIONS
#
state                     
autosave                  0
rewind                    0
rewind_capacity           100
playback                  
record                    
exit_after_playback       0
mngwrite                  
aviwrite                  
wavwrite                  
snapname                  %g/%i
snapsize                  auto
snapview                  auto
snapbilinear              1
statename                 %g
burnin                    0

#
# CORE PERFORMANCE OPTIONS
#
autoframeskip             0
frameskip                 0
seconds_to_run            0
throttle                  1
sleep                     1
speed                     1.0
refreshspeed              0
lowlatency                0

#
# CORE RENDER OPTIONS
#
keepaspect                1
unevenstretch             1
unevenstretchx            0
unevenstretchy            0
autostretchxy             0
intoverscan               0
intscalex                 0
intscaley                 0

#
# CORE ROTATION OPTIONS
#
rotate                    1
ror                       0
rol                       0
autoror                   0
autorol                   0
flipx                     0
flipy                     0

#
# CORE ARTWORK OPTIONS
#
artwork_crop              0
fallback_artwork          
override_artwork          

#
# CORE SCREEN OPTIONS
#
brightness                1.0
contrast                  1.0
gamma                     1.0
pause_brightness          0.65
effect                    none

#
# CORE VECTOR OPTIONS
#
beam_width_min            1.0
beam_width_max            1.0
beam_dot_size             1.0
beam_intensity_weight     0
flicker                   0

#
# CORE SOUND OPTIONS
#
samplerate                48000
samples                   1
volume                    0
compressor                1
speaker_report            0

#
# CORE INPUT OPTIONS
#
coin_lockout              1
ctrlr                     
mouse                     0
joystick                  1
lightgun                  0
multikeyboard             0
multimouse                0
steadykey                 0
ui_active                 0
offscreen_reload          0
joystick_map              auto
joystick_deadzone         0.15
joystick_saturation       0.85
joystick_threshold        0.3
natural                   0
joystick_contradictory    0
coin_impulse              0

#
# CORE INPUT AUTOMATIC ENABLE OPTIONS
#
paddle_device             keyboard
adstick_device            keyboard
pedal_device              keyboard
dial_device               keyboard
trackball_device          keyboard
lightgun_device           keyboard
positional_device         keyboard
mouse_device              mouse

#
# CORE DEBUGGING OPTIONS
#
verbose                   1
log                       0
oslog                     0
debug                     0
update_in_pause           0
debugscript               
debuglog                  0

#
# CORE COMM OPTIONS
#
comm_localhost            0.0.0.0
comm_localport            15112
comm_remotehost           127.0.0.1
comm_remoteport           15112
comm_framesync            0

#
# CORE MISC OPTIONS
#
drc                       1
drc_use_c                 0
drc_log_uml               0
drc_log_native            0
bios                      
cheat                     0
skip_gameinfo             0
uifont                    default
ui                        cabinet
ramsize                   
confirm_quit              1
ui_mouse                  1
language                  
nvram_save                1

#
# SCRIPTING OPTIONS
#
autoboot_command          
autoboot_delay            0
autoboot_script           
console                   0
plugins                   1
plugin                    
noplugin                  

#
# HTTP SERVER OPTIONS
#
http                      0
http_port                 8080
http_root                 web

#
# OSD INPUT MAPPING OPTIONS
#
uimodekey                 auto
controller_map            none
background_input          0

#
# OSD FONT OPTIONS
#
uifontprovider            auto

#
# OSD OUTPUT OPTIONS
#
output                    auto

#
# OSD INPUT OPTIONS
#
keyboardprovider          auto
mouseprovider             auto
lightgunprovider          auto
joystickprovider          auto

#
# OSD DEBUGGING OPTIONS
#
debugger                  auto
debugger_port             23946
debugger_font             auto
debugger_font_size        0
watchdog                  0

#
# OSD PERFORMANCE OPTIONS
#
numprocessors             auto
bench                     0

#
# OSD VIDEO OPTIONS
#
video                     auto
numscreens                1
window                    0
maximize                  1
waitvsync                 0
syncrefresh               0
monitorprovider           auto

#
# OSD PER-WINDOW VIDEO OPTIONS
#
screen                    auto
aspect                    auto
resolution                auto
view                      auto
screen0                   auto
aspect0                   auto
resolution0               auto
view0                     auto
screen1                   auto
aspect1                   auto
resolution1               auto
view1                     auto
screen2                   auto
aspect2                   auto
resolution2               auto
view2                     auto
screen3                   auto
aspect3                   auto
resolution3               auto
view3                     auto

#
# OSD FULL SCREEN OPTIONS
#
switchres                 0

#
# OSD ACCELERATED VIDEO OPTIONS
#
filter                    1
prescale                  1

#
# OpenGL-SPECIFIC OPTIONS
#
gl_forcepow2texture       0
gl_notexturerect          0
gl_vbo                    1
gl_pbo                    1
gl_glsl                   0
gl_glsl_filter            1
glsl_shader_mame0         none
glsl_shader_mame1         none
glsl_shader_mame2         none
glsl_shader_mame3         none
glsl_shader_mame4         none
glsl_shader_mame5         none
glsl_shader_mame6         none
glsl_shader_mame7         none
glsl_shader_mame8         none
glsl_shader_mame9         none
glsl_shader_screen0       none
glsl_shader_screen1       none
glsl_shader_screen2       none
glsl_shader_screen3       none
glsl_shader_screen4       none
glsl_shader_screen5       none
glsl_shader_screen6       none
glsl_shader_screen7       none
glsl_shader_screen8       none
glsl_shader_screen9       none

#
# OSD SOUND OPTIONS
#
sound                     auto
audio_latency             2

#
# PORTAUDIO OPTIONS
#
pa_api                    none
pa_device                 none
pa_latency                0

#
# OSD MIDI OPTIONS
#
midiprovider              auto

#
# OSD EMULATED NETWORKING OPTIONS
#
networkprovider           auto

#
# BGFX POST-PROCESSING OPTIONS
#
bgfx_path                 bgfx
bgfx_backend              auto
bgfx_debug                0
bgfx_screen_chains        
bgfx_shadow_mask          slot-mask.png
bgfx_lut                  lut-default.png
bgfx_avi_name             auto

#
# SDL PERFORMANCE OPTIONS
#
sdlvideofps               0

#
# SDL VIDEO OPTIONS
#
centerh                   1
centerv                   1
scalemode                 none

#
# SDL FULL SCREEN OPTIONS
#
useallheads               0
attach_window             

#
# SDL KEYBOARD MAPPING
#
keymap                    0
keymap_file               keymap.dat

#
# SDL JOYSTICK MAPPING
#
sixaxis                   0

#
# SDL LIGHTGUN MAPPING
#
lightgun_index1           auto
lightgun_index2           auto
lightgun_index3           auto
lightgun_index4           auto
lightgun_index5           auto
lightgun_index6           auto
lightgun_index7           auto
lightgun_index8           auto

#
# SDL LOW-LEVEL DRIVER OPTIONS
#
videodriver               auto
renderdriver              auto
audiodriver               auto
gl_lib                    auto
```

### Emulated system/software

Macintosh 128k(mac128k)

### Incorrect behaviour

When booting the mac128k with compatible Apple system software disks, the mame process will become frozen causing 100% cpu usage of whichever of my cpu cores it happened to be running on.

I have to kill it with kill -9.  A normal kill will have no effect.

### Expected behaviour

I expect mame to not become frozen in an infinite loop.

### Steps to reproduce

1. I started mame with the following command: 
`mame mac128k sys11`
2. It became frozen while the emulated floppy disk sounds were playing.  It was clear exactly when the freeze had occurred because the floppy sound cut suddenly and completely with a faint clicking sound.

### Additional details

I found that I can get things going again by running mame within gdb.

*I have compiled mame 0.259 from source with these options: DEBUG=1 VERBOSE=1 SYMBOLS=1 OPTIMIZE=0

1. Start gdb with the following command:
`gdb --args ./mamed mac512k sys11`
2. Once gdb has finished reading symbols and the (gdb) prompt appears enter the "run" command to start mame.
3. Press a key to move past the initial title/banner screen.
4. Wait for the initial mac boot up screens to go by.
5. Listen as the emulated floppy sounds play.
6. Once the emulated floppy sound cuts out with a click, the freeze has occurred.
7. Use alt-tab to switch to the terminal where mame is running under gdb.
8. Press ctrl-c to interrupt execution.
9. Enter the "finish" command to finish the current stack frame.
10. Keep running the "finish" command until you are in device_scheduler::execute_timers.
- If you go too far you might end up stuck in the loop again.  Go back to step 8.
11. Run the "next" command until you are at line 956 of src/emu/schedule.cpp
`if (!m_callback_timer_modified)`
- Again, if you go too far you might end up stuck in the loop again.  Go back to step 8.
12. Enter the following command to set m_callback_timer_modified to false
`set m_callback_timer_modified = false`
13. Enter the "continue" command to resume execution.
14. Things should be proceeding normally now.  Use alt-tab to switch back the emulated mac window which should be booting as normal.

*Note: I was also seeing this in mame 0.258

**Comment by rb6502** [comment]
Seems to be due to the 400K drive; sys11 works with the mac512ke and macplus drivers.  Poking @galibert ...

**Comment by Davidian1024** [comment]
It's definitely not limited to 400K drives.  When I boot the mac128ke with sys608 it locks up the same way as the mac128k with sys11 does.  That system has two 800K/400K drives and the 608 disk images are 800K.

There's actually alot of scenarios where it will get locked up this way.  So far, I'm always able to get things going again with that m_callback_timer_modified trick.

And it's not limited to booting system software.  Once I've gotten it past the initial freeze, I've tried things like initializing disks, loading disks with software on them to run games, etc.  It seems that whenever disk access happens there is a chance this freeze will happen.

There's another issue I hadn't mentioned yet as well.  I wasn't sure if it's completely separate or not.  But, more and more I'm thinking it's actually the same thing.  While the disk access is happening I hear a fair amount of clicking.  Sometimes I think it actually gets frozen briefly.  What I think is happening in these cases is that it's getting caught in the same loop, but only for a brief moment.  Enough to cause audio issues.

**Comment by rb6502** [comment]
Poking @galibert again, because this sounds like I should remove the Mac support level of bad.

**Comment by Davidian1024** [comment]
Are you saying you're thinking you should change the Mac support level from bad to what? not supported?

**Comment by rb6502** [comment]
I should've used more punctuation.  "like 'I should remove the Mac support' level of bad" was the intent.

Checking further it's fine in 0.258 so it's likely the changes that went in for hard-sectored floppy support.

**Comment by galibert** [comment]
Noted, will have a look this week.

**Comment by Davidian1024** [comment]
I initially encountered these freezes under 0.258.  I only started really digging into it once 0.259 was released.
I just tested the two following scenarios with the 0.258 binaries I still have to confirm.
`mame mac128k sys11`
`mame mac512ke sys608`
both of these always freeze for me at the little happy mac icon screen.

**Comment by rb6502** [comment]
Freeze as in MAME locking up or freeze as in not loading further?  I don't see either of those problems on 0.258.  Try deleting the contents of your diff/ folder in case a write to those disks on 0.259 broke something.

**Comment by Davidian1024** [comment]
MAME freezes.  It stops responding to any input.  ScollLock no longer works to enable the UI.  The thread it's running on jumps to 100% cpu usage while it's stuck like this.  I have to kill it with kill -9.

Still happens after clearing diff/ just a bit ago.
I did that with both the 0.258 and 0.259 binaries I have.

FWIW, I didn't see any diffs in there for those disk images.  And I was running the 0.259 binary I built in a different directory so it was going off of a different diff/ directory.

I think I might see the fix you're talking about, gonna try building a new binary with that and see if it helps.

**Comment by rb6502** [comment]
I vaguely remember there's some issue with Ubuntu's default GCC options making it generate bad code for MAME, but I don't remember the details.  I know the freeze is a 100% thing on shipped 0.259 but I'm not seeing it at all on 0.258 or top-of-tree.

**Comment by Davidian1024** [comment]
Ok so, I built a top-of-tree binary with SOURCES=src/mame/apple/mac128.cpp
MAME now no longer gets completely frozen like it had been.

I am still getting the popping and clicking sounds as it accesses a floppy.  And when this happens I can now say for sure that MAME is getting frozen briefly.  I can tell because I'm pulling up the UI menu and moving up and down within it as the mac is booting up behind it.  Whenever these click freezes occur, the MAME UI menu is frozen during that moment.

**Comment by Davidian1024** [comment]
Ok so, I definitely made one mistake in what I reported earlier.  It turns out that the version 0.258 binary I've been talking about above is built from a clone I took somewhere between the time 0.258 was released and 0.259 was released.  It was just a top-of-tree clone from that time.

I cloned the 0.258 branch separately and built it from that directory.  I don't get the same freezes I had been reporting with the 0.259 version.  I do get the popping/clicking audio with some momentary freezing/hesitation, but it doesn't get stuck forever.

Sorry for the confusion.
